musmart/melody/contour/step_contour.py

```diff
@@ -0,0 +1,247 @@
+        Float value representing the global variation of the step contour
+
+    Raises
+    ------
+    ValueError
+        If the length of the step contour is not 64
+
+    Examples
+    --------
+    Calculate the global variation of a step contour
+    >>> contour = [60, 60, 60, 60, 62, 62, 62, 62, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63,
+    ...           63, 63, 63, 63, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65,
+    ...           72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72,
+    ...           72, 72, 72, 72, 72, 72, 72, 72, 72, 72]
+    >>> step_contour_global_variation(contour)
+    4.463392767839281
+
+    Create step contour by adding lists
+    >>> descending = [72] * 16 + [69] * 16 + [65] * 16 + [60] * 16
+    >>> step_contour_global_variation(descending)
+    4.5
+
+    Calculate global variation for a constant melody
+    >>> constant = [60] * 64
+    >>> step_contour_global_variation(constant)
+    0.0
+
+    Throws a ValueError when contour length is not 64
+    >>> step_contour_global_variation([60, 62, 63]) # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+      ...
+    ValueError: The length of the step contour must be 64.
+    """
+    if len(contour) != 64:
+        raise ValueError("The length of the step contour must be 64.")
+
+    return float(np.std(contour))
+
+
+def step_contour_global_direction(contour: list[int]) -> float:
+    """Calculates the global direction of a step contour by correlating the step
+    contour with a vector of [1, 2, 3, ..., 64].
+
+    Parameters
+    ----------
+    contour : list of int
+        List of length 64 containing the step contour of a melody
+
+    Returns
+    -------
+    float
+        Float value representing the global direction of the step contour
+        Returns 0.0 if the contour is flat
+
+    Raises
+    ------
+    ValueError
+        If the length of the step contour is not 64
+
+    Examples
+    --------
+    Calculate the global direction of a step contour - this example is clearly upwards
+    >>> contour = [60, 60, 60, 60, 62, 62, 62, 62, 63, 63, 63, 63, 63, 63, 63, 63, 63, 63,
+    ...           63, 63, 63, 63, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65,
+    ...           72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72, 72,
+    ...           72, 72, 72, 72, 72, 72, 72, 72, 72, 72]
+    >>> step_contour_global_direction(contour)
+    0.9316021128080264
+
+    Calculate the global direction of a step contour - this example is clearly downwards
+    >>> descending = [72] * 16 + [69] * 16 + [65] * 16 + [60] * 16
+    >>> step_contour_global_direction(descending)
+    -0.9623679323746963
+
+    Flat contour
+    >>> flat = [60] * 64
+    >>> step_contour_global_direction(flat)
+    0.0
+    """
+    if len(contour) != 64:
+        raise ValueError("The length of the step contour must be 64.")
+
+    corr = np.corrcoef(contour, np.arange(64))[0, 1]
+    if np.isnan(corr):
```

Comment:
You're right. This implementation was too simplistic, aimed at dealing with the case where the input list contains all the same values `[60, 60, 60, ..., 60]` - this by definition causes a division by zero problem, and `np.corrcoef` returns `NaN` as a result. I've fixed it now with a case to catch repeated values in input, and if an input is invalid, it will instead return `None`, rather than a value that could be construed as a correlation coefficient.